tests/System.Text.JsonLab.Tests/ReadFromPipeTests.cs

```diff
@@ -0,0 +1,86 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.IO.Pipelines;
+using System.Text.JsonLab.Tests.Resources;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.Text.JsonLab.Tests
+{
+    public class ReadFromPipeTests : IDisposable
+    {
+        public ReadFromPipeTests()
+        {
+            string jsonString = TestJson.Json400KB;
+            _dataUtf8 = Encoding.UTF8.GetBytes(jsonString);
+            _written = 0;
+            _pipe = new Pipe();
+        }
+
+        public void Dispose()
+        {
+            _pipe.Writer.Complete();
+            _pipe.Reader.Complete();
+        }
+
+        private readonly Pipe _pipe;
+        private readonly byte[] _dataUtf8;
+        private int _written;
+
+        [Fact]
+        public void ReadFromPipe()
+        {
+            var taskReader = Task.Run(async () =>
+            {
+                SequencePosition position = default;
+                JsonReaderState state = default;
+                while (true)
+                {
+                    ReadResult result = await _pipe.Reader.ReadAsync();
+                    ReadOnlySequence<byte> data = result.Buffer;
+                    bool isFinalBlock = data.Length == 0;   // TODO: Fix isFinalBlock condition, if Writer is done instead, isLastSegment?
```

Comment:
This isn't the right condition to set isFinalBlock. It needs to be based on whether Writer has completed or not.

I was thinking of using IsLastSegment (making it a field), but I can only set IsLastSegment after writing the last segment, and that could lead to a race where the reader reads the data before I set IsLastSegment to true.

Maybe something related to `_pipe.Reader.OnWriterCompleted()` can tell me if isFinalBlock should be set or not.

Any suggestions?

Reply:
That condition results in an extra (unnecessary) call to Read with a knowingly empty json payload. It also doesn't catch invalid json scenarios like these:
Single segment : `1234.` <= should throw since the number ends with a decimal point.

We don't know isFinalBlock is true, after the decimal we return false. So, data.Length is never == 0, and we never get isFinalBlock is true.

Should it be as follows instead?
`bool isFinalBlock = result.IsCompleted;`

Doing so solves the correctness issue that I stated above.